[lesterlimjj] iP

lesterlimjj opened the pull request:
# Owen: a friend who remembers
>If you turn into a table
>You must at the height of my elbows
>Odd stubby leg makes you wobbly
>You're the only perfect table for me
> \- lyrics from world.search(you); by mili

Tell Owen everything and he will remember it for you! It's
- user friendly
- ***Faster Than Light*** (An old reference hahaha)
- for those who enjoy typing

**Shut up and take my money!!!** 
~~Sure,~~ I mean ahem...*its free*, you have to,
1. download [Owen's jar](https://github.com/lesterlimjj/ip/releases/tag/A-Jar)
2. double-click it or run java -jar Owen.jar in the directory Owen is downloaded
3. add thee tasks
4. Your done! Owen does the rest! 🥳 

Owen's features:
- [x] Managing your tasks (todos, events and deadline)
- [x] Automatic saving of tasks
- [ ] GUI implementation (coming soon!) 

Wanna figure out how Owen works? Here is a code snippet of the `run` method.
```java
while (!isExiting) {
    try {
        String input = scanner.nextLine();
        ui.showSeparator();
        Command command = Parser.parse(input);
        command.execute(ui, storage, tasks);
        isExiting = command.isBye();

    } catch (OwenException exception) {
        ui.showMessage(exception.getMessage());
    }
```
You can checkout the codebase [here](https://github.com/lesterlimjj/ip).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

lesterlimjj committed:
renamed Duke.java

lesterlimjj committed:
Added greet and exit

lesterlimjj committed:
add echo feature

lesterlimjj committed:
add support for choosing between commands

lesterlimjj committed:
add list and add to list features

lesterlimjj committed:
add Task class

lesterlimjj committed:
Updated main to use Task class

lesterlimjj committed:
update list feature to show task status

lesterlimjj committed:
add mark as done feature

lesterlimjj committed:
Add mark as not done feature

lesterlimjj committed:
add Todo class

lesterlimjj committed:
add deadline class

lesterlimjj committed:
add Event class

lesterlimjj committed:
add support for adding Todo, Deadline, Event to tasklist

lesterlimjj committed:
fixed bugs for echo and list feature and corrected formatting issues when displaying text

lesterlimjj committed:
added semi-automated testing of chatbot

lesterlimjj committed:
Added Exception class

lesterlimjj committed:
Added exception handling for invalid input

lesterlimjj committed:
Add exception handling for unknown user inputs and invalid inputs for todo, deadline and event

lesterlimjj committed:
add invalid input handling to mark and unmark feature

lesterlimjj committed:
add delete feature

lesterlimjj committed:
Changed any mentions of Duke to Owen

lesterlimjj committed:
Updated error handling for deadline and event to be more stringent

lesterlimjj committed:
fix bug for Todo

lesterlimjj committed:
Update error handling for mark, unmark and delete to cover wider range of errors

lesterlimjj committed:
Updated semi automated testing to check for new error handling

lesterlimjj committed:
added separators to text for neatness

lesterlimjj committed:
updated expected result from test to have dividors

lesterlimjj committed:
Add loadTasklist function on startup

The chatbot has no ability to retain previously written tasks by
the user.

It would enhance the functionality of the chatbot to be able to
remember tasks so it needs to be able to load data.

Let's:
Use a text file to load Todo, Deadline, and Event items when the
chatbot starts, with the use of the java.nio.file library.
Ensure the chatbot creates a data folder and an empty text file
if they do not exist on startup.

If the chatbot is able to load data from a text file, we can alter
this text file to include user data.

lesterlimjj committed:
Add convertToDataFormat function to Task and its children

We have no convenient way of producing the required data format for
a Task class and its children.

We need a function to streamline the process of getting the string
to be written to text file.

Let's include the function in Task and use polymorphism to create
the child's own convertToDataFormat function

Using polymorphism reduces the amount of checks needed for what
format should be used depending on the class.

There are common fields in the data between Task and its children
so polymorphism is suitable to be used.

lesterlimjj committed:
Add overwriteTasklist and appendToTasklist functions

we have no functions to modify the data in the text file.

we need a way to save the user's changes to the tasklist to
complement the load tasklist functionality.

Lets,
*create a appendToTasklist function that append a new line of data
to the text file.
*create an overwriteTasklist function that loop through the tasklist
arraylist to overwrite the text file data by writing to it line by
line.

We use appendToTasklist when adding a new Task to the Text for
efficiency (since it directly adds to the list) while overwrite
is only used for modifying existing data (e.g. mark and delete).

lesterlimjj committed:
Merge branch 'branch-level-7'

lesterlimjj committed:
Add processLocalDateTime function

Chatbot takes in dates as Strings and simply repeats them.

Chatbot usability would improve if it can handle a range of
datetime formats from user

Let's
* add processLocalDateTime function to parse a set of defined
datetime formats
* ensure processLocalDateTime can handle both date and datetime by
padding date with default time value

Doing this will allow the user more flexibility in the type of
date inputs given.

lesterlimjj committed:
Update Deadline and Event class to handle LocalDateTime

Using String for dates makes error handling and formatting much
more difficult.

Let's update Deadline and Event to use LocalDateTime instead of
String for any dates and update the toString to use a more readable
date format.

Should allow users to at a glance tell what datetime it is.

lesterlimjj committed:
Updated processLocalDateTime to no longer handle LocalDate

processLocalDateTime was taking LocalDate and convert it to
LocalDateTime with a default value of 12am.

It should be removed since it can confuse the user with a value he
did not input.

lesterlimjj committed:
Reorder localDateTimePatterns

The order made it so that the second format would never be checked.

Reduces the amount of formats a user can input.

Lets swap the order so that the second format is now checked first.

lesterlimjj committed:
Merge branch 'branch-Level-8'

lesterlimjj committed:
Extract UI-related functions into a separate UI class

UI-related functions (e.g., greetings, task list display) were
previously embedded in the main method.

Including UI functions in the main method made the code less
modular and harder to maintain.

Let's extract all UI-related functions into a new UI class. This
class encapsulates responsibilities such as displaying greetings,
task lists, separators, and task counts.

Separating UI logic improves code readability and maintainability.

lesterlimjj committed:
Move processLocalDateTime method to Event and Deadline class

The processLocalDateTime method is located in main.

Placing processLocalDateTime in main caused repetitive calls
since whenever event or deadline object was created, String had
to be converted to LocalDateTime

Let's move the processLocalDateTime method into the Event and
Deadline class, as it directly pertains to parsing the dates
required by only these two classes.

Reduces the amount of times this function is called

lesterlimjj committed:
Extract Storage class from main

File handling logic for loading and saving tasks is located in the
main class.

Having file operations in main creates unnecessary coupling, making
 the code harder to maintain.

Let's extract all file handling responsibilities into a dedicated
Storage class. This includes: loading tasklist, overwriting tasklist
data (when data is modified) and appending new entry to tasklist.

A separate Storage class improves maintainability by allowing
storage logic to be modified independently of other components.

lesterlimjj committed:
Extract TaskList class from main

The main class handles task management operations, such
as adding, deleting, and updating tasks.

Having task management logic in main makes the code harder to scale
and maintain.

Let's extract all task management operations into a dedicated
TaskList class. This class will handle adding/deleting of tasks,
retrieving the list of tasks and any task updating (e.g. mark).

This allows our tasklist to be more modular, easier to test, and
maintainable.

lesterlimjj committed:
Extract abstract command class and all children commands from main

All user commands are handled in main and logic for processing
command intertwined with other components.

This causes reduced flexibility and increasing complexity for user
commands.

Let's:
* Introduce an abstract Command class to represent general
commands. It has an abstract execute method and isBye method to
handle chatbot exit commands.
* add concrete implementation of user command based on abstract
Command class

This way, adding and updating user commands is easier and more
manageable

lesterlimjj committed:
Extract parsing logic from main into dedicated Parser class

The main class contains logic for parsing user input and
handling various commands.

This makes the main class overly complex and tightly coupled with
the parsing process.

Let's
* extract the parsing logic into a new Parser class. The class
processes user input, validates it, and constructs appropriate
command objects.
* the parse class will also date parsing and handle exception for
commands and tasks.

Doing this will ensure a clear separation of concern for parsing
and exception handling.

lesterlimjj committed:
Remove date parsing functions in Deadline and Event class

Date parsing is already handled by Parser class so these functions
are redundant.

lesterlimjj committed:
Refactor Owen to delegate responsibilities to other components

The Owen class handles initialization, user interaction, command
parsing, and task management, leading to a bloated and less modular
design.

Centralizing too many responsibilities in the Owen class makes the
code harder to read, maintain, and extend.

Let's refactor the Owen class to have parser, tasklist and storage
components.

This approach simplifies the Owen class and adheres to the single
responsibility principle. Additionally, each component focuses on a
single task.

lesterlimjj committed:
Add error checks for mark and unmark to Parser

lesterlimjj committed:
Merge branch 'branch-A-MoreOOP'

lesterlimjj committed:
Organize classes into packages to improve project structure

All classes are located in the root directory, leading to a
cluttered and unorganized project structure.

The lack of structure makes it harder to maintain the codebase as
the project grows.

Let's organize the classes into the following packages:
command: Contains all command-related classes (e.g. Command,
AddDeadlineCommand, etc.).
task: contains task-related classes (e.g. Todo, Deadline, Event).
parser: contains parsing related class
storage: contains file operation storage class
ui: contains user interaction classes
exception: contains exception related class

Grouping related classes into packages improves modularity and
readability.

lesterlimjj committed:
Merge branch 'branch-A-packages'

lesterlimjj committed:
Merge branch 'add-gradle-support'

lesterlimjj committed:
Update build.grade mainClass to Owen

lesterlimjj committed:
Add junit tests for Parser and Storage

Testing for parser and storage required a lot of manual analysis.

Manual analysis is time consuming and prone to human error.

Let's use Junit to create automated tests of Storage and Parser
functions.

Allows for easy checking if new changes break any pre-existing
functionality.

lesterlimjj committed:
Merge branch 'branch-A-JUnit'

lesterlimjj committed:
configure shadowJar to produce correct Jar file

lesterlimjj committed:
Merge branch 'branch-A-Jar'

lesterlimjj committed:
Update code files to follow Java SE coding standards

Codebase does not fully adhere to Java SE coding standards,
leading to inconsistencies in formatting, naming conventions,
and readability.

Non-compliance with Java SE coding standards makes the code harder
to read, maintain, and collaborate on.

Let's update the coding standards based on the provided
java coding standards.

Adhering to Java SE standards promotes clean and maintainable code

lesterlimjj committed:
Add Javadocs to all classes and methods

All methods and classes in the codebase lack proper Javadoc
comments.

Future contributors will have a harder time understanding the
classes and methods.

Let's add Javadoc to all classes and methods.

Using Javadoc ensures consistency and allows tools to generate
external documentation.

This makes it easier to share and understand the project's structure
 and functionality.

lesterlimjj committed:
Add find command to search for tasks

The application lacks a feature to search tasks based on
words.

It might become hard for user to locate a specific task when the
list becomes really large.

Let's create a command that searches the task list based on a word
or phrase, returning all task with the word or phrase.

This approach ensures simplicity and clarity, leveraging string
matching to identify relevant tasks.

lesterlimjj committed:
Merge branch 'branch-A-CodingStandard'

lesterlimjj committed:
Merge branch 'branch-A-JavaDoc' and resolve conflicts

Master merged with coding standard branch, then JavaDoc branch.

Both branches cover very similar areas in terms of code edits

# Conflicts:
#	src/main/java/Owen.java
#	src/main/java/command/DeleteCommand.java
#	src/main/java/parser/Parser.java
#	src/main/java/storage/Storage.java
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java
#	src/main/java/ui/Ui.java

Let's resolve this by keeping both changes from A-JavaDoc and
A-CodingStandards.

We need both Java coding standards and JavaDoc for codebase
scalability.

lesterlimjj committed:
Merge branch 'branch-Level-9' and resolve conflicts

Master was using an older version of command switch case as
compared to Level 9.

# Conflicts:
#	src/main/java/parser/Parser.java

Let's resolve this conflict by taking the branch Level 9 version of
switch case.

The newest version of command switch case has find command.

lesterlimjj committed:
Add checkstyle

There is no automated way of checking for Java coding standards,
aside from manual checks.

Manual checks are inefficient and error-prone.

Let's add Ab3's checkstyle to check for Java coding standards. This
includes:
* adding checkstyle configuration files.
* adding checkstyle-IDEA plugin.

This allows for an automated way of testing as well as being alerted
to style violations while coding, ensuring higher code quality in a
more efficient manner.

lesterlimjj committed:
Merge branch 'branch-A-CheckStyle'

lesterlimjj committed:
Add JavaFx UI Components

Application is fully text-based.

This limits the interactivity and user experience.

Let's use JavaFx to introduce GUI. This includes:
*a DialogBox component to replace user and chatbot text.
*a Window component which contains DialogBoxs and a text field with
a submit button.

These will appropriately serve as the UI components without
requiring drastic changes to the logic.

lesterlimjj committed:
Add GuiController

UI components has no way to send data to the logic in Owen class.

Unable to create commands from user input.

Let's create a GuiController class that manages the updating of
Gui elements. It does the following:
*update the window with new user and chatbot dialog boxes.
*passes user input to Owen for evaluation.

Keep the code neat and scalable by keeping all UI management in a
class, rather than across multiple UI components.

lesterlimjj committed:
Update Commands and its subclasses execute function

Execute function uses old UI class for text-based UI.

This is incompatible with JavaFX UI Component.

Let's updates them to use GuiController to add user and chatbot
dialog boxs.

This ensures that the GUI is appropriately updated after a
command is executed.

lesterlimjj committed:
Add Error Command class

Thrown exception are printed as text.

GUI need a way to show these exception.

Let's add an Error Command class that displays the error message
in these exception.

Ensures that every exception thrown can be displayed

lesterlimjj committed:
Merge branch 'branch-Level-10'

xinnnyeee commented on src/main/java/owen/Owen.java:
I like how each field and method is so well-documented!

xinnnyeee commented on src/main/java/owen/Owen.java:
```diff
@@ -0,0 +1,83 @@
+            if (command.isBye()) {
+                System.exit(0);
+            }
+
```
very comprehensive error handling! and I like how you put ur exception handling in a separate class

xinnnyeee commented on src/main/java/owen/command/Command.java:
interesting way to handles the commands (ie breaking into different classes)! I think I learned something new here! if you're free to respond, can share with me how you came up w this and how well you feel its working?

lesterlimjj replied:
oh for the commands, it was an optional feature suggestion in week 3 more OOP. For more OOP, they suggested referring to [address book 2](https://github.com/se-edu/addressbook-level2/blob/master/src/seedu/addressbook/commands) structure which I found quite logical. Hence, my code is inspired by their implementation.

xinnnyeee commented on src/test/java/owen/parser/ParserTest.java:
parser testing very comprehensive!

xinnnyeee reviewed (commented):
In general LGTM! Jiayouu

xuxuezhou commented:
The code standard LGTM, good job!

xuxuezhou commented on src/main/java/owen/ui/DialogBox.java:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Flips the dialog box such that the ImageView is on the left and text on the right.
+     */
+    private void flip() {
+        ObservableList<Node> tmp = FXCollections.observableArrayList(this.getChildren());
+        Collections.reverse(tmp);
+        getChildren().setAll(tmp);
+        setAlignment(Pos.TOP_LEFT);
+    }
```
I really like how the flip() method elegantly handles the layout switch between user and bot dialogs—great implementation!

codebreaker64 commented on src/main/java/owen/Owen.java:
```diff
@@ -0,0 +1,83 @@
+        ui.setOwen(this);
+        tasks = new TaskList();
+        storage = new Storage();
+        storage.loadTasklistData(tasks);
```
Should be loadTask"L"istData since you are using TaskList for the class.

codebreaker64 commented on src/main/java/owen/command/AddDeadlineCommand.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Represents a command to add a deadline to the task list.
+ */
+public class AddDeadlineCommand extends Command {
```
The class should be a noun instead of a verb, like "DeadlineCommand"

codebreaker64 commented on src/main/java/owen/command/AddEventCommand.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Represents a command to add an event to the tasklist.
+ */
+public class AddEventCommand extends Command {
```
Should be changed to "EventCommand"

codebreaker64 commented on src/main/java/owen/command/AddTodoCommand.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Represents a command to add a todo to the task list.
+ */
+public class AddTodoCommand extends Command {
```
Should be changed to TodoCommand

codebreaker64 reviewed (commented):
Overall, good naming quality except for some class can change from verb to noun.

lesterlimjj committed:
Change method names with tasklist to taskList

Method names used tasklist but the class being referred to is
named TaskList.

This creates inconsistency in the naming convention.

Let's change all tasklist in method names to taskList.

This makes it more evident that our naming convention for method
names adheres to camelcase.

Note that junit test case methods are also affected by this change.

lesterlimjj committed:
Replace low level statements in execute method in Command classes

Execute function perform string formatting and concatenation
operations in Command classes.

This violates SLAP.

Let's create a formatResponse method in GuiController to replace
these low level statements.

This ensures that level of abstraction is the same in the execute
method.

lesterlimjj committed:
Extract methods from parse method

Every switch case statement for a command type has the logic of
processing that command contained in the switch case statement.

The lines of code for Parse exceeded 30 lines, making it hard to
understand and maintain.

Let's extract out a process method for each command type.

This improves readability and ensures that SLAP is easily
maintained in the switch case statements.

lesterlimjj committed:
Extract methods from loadTaskListData method

loadTaskListData reads the text file data, parses the data into
tasks and adds the tasks to the task list. It also creates the data
file and directory if they does not exist.

The logic for operations are all inside of loadTaskListData, making
the lines of code over 30 and violating SLAP in certain areas.

Let's extract out these operations into methods. In particular, we
have:
*a method for creating the data file and directory if they does
not exist
*a method for splitting and trimming the data
*a method for creating and adding todo from the data
*a method for creating and adding deadline from the data
*a method for creating and adding event from the data

This improve code readability and ensures that it is easier to
maintain.

lesterlimjj committed:
Add assertions to various methods

There is no usage of assertions in the code.

Assertions allow us to validate assumptions made during runtime,
ensuring the code behaves as expected.

Let's add assertions to our Command subclasses, Task subclasses and
exception class. In particular, we can perform assertions on the
fields in these classes to ensure they are not null or have
unexpected values during runtime.

This ensures that much fewer bugs go undetected as the program is
developed.

lesterlimjj committed:
Merge pull request #1 from lesterlimjj/branch-A-CodeQuality

Merge Branch-A-CodeQuality into master

lesterlimjj committed:
Merge branch 'master' of https://github.com/lesterlimjj/ip

lesterlimjj committed:
Merge branch 'master' into branch-A-Assertions

lesterlimjj committed:
Merge pull request #2 from lesterlimjj/branch-A-Assertions

Merge branch-A-Assertions into master branch

lesterlimjj committed:
Merge branch 'master' of https://github.com/lesterlimjj/ip

lesterlimjj committed:
Add tagging feature

The program has no way of tagging a task.

Tagging can provide useful new information to the user.

Let's,
* create a addTagCommand
* add tags as a ArrayList<String> field to Task
* update toString and convertToDataFormat to account for tags
* update parser to understand addTagCommand
* update loadTaskListData to account for tags

Through inheriting Task, Todo, Event and Deadline will have tag
field and any tag-related methods.

lesterlimjj committed:
Merge branch 'branch-C-Tagging'

lesterlimjj committed:
Update Javadocs comments format

Issue tracker indicates that some Javadocs comments start with
the wrong format and are missing punctations.

This creates inconsistency and lowers code quality.

Let's
*correct comments to be in the form of Adds, Returns, etc
*add punctuations where appropriate

This will make our code align with the given Java coding standard.

lesterlimjj committed:
Improve error handling for event, deadline and tags

Event and deadline error handling do not check for empty inputs.
Tags have no error handling.

This can cause unexpected responses to the user when handling
errors.

Let's
* add error handling for tags that cover wrong number of command
parameters and empty inputs
* add error handling for empty date strings and start date after
end date

This ensures the error handling is more comprehensive.

lesterlimjj committed:
Add delay to exit when bye command is given

Application closes immediately and bye message is not shown.

This worsens user experience since it can be jarring for
application to close immediately without a response.

Let's
* add a 1 second delay from JavaFx when exiting the application

This ensures that there is time to show the bye message before
exiting the application.

lesterlimjj committed:
Add User Guide for Owen chatbot

There is no user guide for Owen and no help command as well.

New users will have a difficult time learning how to use the
product.

Let's,
* add a README document that specifies the commands and their
usage

This ensure that there is a proper guide for new users to follow.

lesterlimjj committed:
Fix typos in README.md

lesterlimjj committed:
Update Parser to remove magic numbers

lesterlimjj committed:
Fix formatting issues for README.md

lesterlimjj committed:
Update build.gradle to build correct jar file

lesterlimjj committed:
Update user and Owen images to be smaller resolution

lesterlimjj committed:
Update exception handling for event and deadline

lesterlimjj committed:
Update Ui image

lesterlimjj committed:
Remove unused variable in Storage method

lesterlimjj committed:
Update Javadoc comments for Parser and Task methods
